pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+)
+
+type v3Helper struct {
+	// etcd interfaces
+	client *clientv3.Client
+
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+func newV3Helper(c *clientv3.Client, codec runtime.Codec, prefix string) *v3Helper {
+	return &v3Helper{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (h *v3Helper) Backends(ctx context.Context) []string {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+		ctx = context.TODO()
+	}
+	resp, err := h.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (h *v3Helper) Codec() runtime.Codec {
+	return h.codec
+}
+
+func (h *v3Helper) Versioner() storage.Versioner {
+	return h.versioner
+}
+
+func (h *v3Helper) Set(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	rev, err := extractRevision(h.versioner, obj)
+	if err != nil {
+		return err
+	}
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// Note that if rev = 0, it means the key doesn't exist. In such case, it's equivalent to create.
+	txnResp, err := h.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", rev),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	// Compare failed
+	if !txnResp.Succeeded {
+		return storage.NewResourceVersionConflictsError(key, rev)
+	}
+
+	if out != nil {
+		// get the response of the put request
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(h.codec, h.versioner, data, out, putResp.Header.Revision)
+	}
+
+	return nil
+}
+
+func (h *v3Helper) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	// if not found:
+	//   if ignoreNotFound = true: set objPtr to be zero value
+	//   else: return not found error
+	if len(getResp.Kvs) == 0 {
+		if ignoreNotFound {
+			return setZeroValue(out)
+		} else {
+			return storage.NewKeyNotFoundError(key, 0)
+		}
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	if version, err := h.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return errors.New("resourceVersion should not be set on objects to be created")
+	}
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// If rev = 0, it means the key doesn't exist.
+	resp, err := h.client.KV.Txn(ctx).If(
```

Comment:
In Set() you call resp: "txnResp". Can you make it consistent (either changing this or the previous one).